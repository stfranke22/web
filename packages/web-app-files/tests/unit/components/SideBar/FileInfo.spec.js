import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import stubs from '@/tests/unit/stubs/index.js'
import { createLocalVue, mount } from '@vue/test-utils'
import FileInfo from '@files/src/components/SideBar/FileInfo.vue'

import GetTextPlugin from 'vue-gettext'

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const filesPersonalRoute = { name: 'files-personal' }
const filesPublicRoute = { name: 'files-public-list' }

function getWrapper(
  route,
  { filename = 'testfile', extension, type = 'file', publicLink = false }
) {
  const mountStubs = { ...stubs, 'oc-button': false }
  const mountOptions = {
    localVue,
    store: createStore({}, filename, extension, type),
    stubs: mountStubs,
    mocks: {
      $route: route,
      publicPage: () => publicLink
    },
    propsData: {
      isContentDisplayed: true
    }
  }
  return mount(FileInfo, mountOptions)
}

const selectors = {
  favoritesButton: '.file_info button.file_info__favorite'
}

describe('FileInfo', () => {
  describe.each([[filesPersonalRoute], [filesPublicRoute]])(
    'FileInfo Components on %s routes',
    route => {
      it.each([
        [{ filename: 'testfile', extension: 'txt', type: 'file' }],
        [{ filename: 'test-file', extension: 'drawio', type: 'file' }],
        [{ filename: 'testfolder', extension: '', type: 'folder' }]
      ])('should show the fileinfo correctly', data => {
        const wrapper = getWrapper(route, data)

        expect(wrapper).toMatchSnapshot()
      })

      it('should call the favorite trigger when button is clicked', async () => {
        const wrapper = getWrapper(route, {
          filename: 'testfile',
          extension: 'jpg'
        })

        const favoriteTrigger = jest.fn()
        wrapper.vm.markFavorite = favoriteTrigger
        const favoritesButton = wrapper.find(selectors.favoritesButton)
        await favoritesButton.trigger('click.native.stop')
        expect(favoriteTrigger).toHaveBeenCalledTimes(1)
      })
    }
  )
})

function getResource({ filename, extension, type }) {
  // convert date from 2 days ago to format 'Mon, 12 Jul 2021 11:04:33 GMT'
  const d = new Date()
  d.setDate(d.getDate() - 2)
  const mdate = d.toGMTString()
  return {
    id: '4',
    fileId: '4',
    icon: type,
    name: type === 'file' ? `${filename}.${extension}` : filename,
    extension: extension,
    path: type === 'file' ? `/${filename}.${extension}` : `/${filename}`,
    type,
    mdate,
    size: '163',
    indicators: [],
    permissions: 'RDNVW',
    starred: false,
    shareTypes: [],
    ownerDisplayName: 'user1',
    ownerId: 'user1',
    canDownload: () => true,
    isReceivedShare: () => true,
    canBeDeleted: () => true,
    canRename: () => true
  }
}

function createStore(state, filename, extension, type = 'file') {
  return new Vuex.Store({
    modules: {
      Files: {
        state: {
          ...state,
          currentFolder: { path: '' }
        },
        namespaced: true,
        getters: {
          highlightedFile: () => {
            return getResource({ filename, extension, type })
          }
        },
        actions: {
          markFavorite: jest.fn()
        }
      }
    },
    getters: {
      capabilities: () => {
        return {
          files: {
            favorites: true
          }
        }
      }
    }
  })
}
