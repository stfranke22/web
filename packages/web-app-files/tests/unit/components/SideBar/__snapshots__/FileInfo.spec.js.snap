// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInfo FileInfo Components on { name: 'files-personal' } routes should show the fileinfo correctly 1`] = `
<div class="file_info">
  <oc-icon-stub name="file" accessiblelabel="" type="span" size="large" variation="passive" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body">
    <h2 tabindex="-1"><span data-test-resource-path="/testfile.txt" data-test-resource-name="testfile.txt" data-test-resource-type="file" class="oc-resource-name oc-text-truncate"><!----><span class="oc-resource-basename">testfile</span><span class="oc-resource-extension">.txt</span></span></h2>
    <div>
      163 B,

      2 days ago
    </div>
  </div> <button aria-label="Click to mark this file as favorite" class="file_info__favorite oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
    <oc-icon-stub name="star" accessiblelabel="" type="span" size="medium" variation="passive" class="oc-star-dimm"></oc-icon-stub>
  </button>
</div>
`;

exports[`FileInfo FileInfo Components on { name: 'files-personal' } routes should show the fileinfo correctly 2`] = `
<div class="file_info">
  <oc-icon-stub name="file" accessiblelabel="" type="span" size="large" variation="passive" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body">
    <h2 tabindex="-1"><span data-test-resource-path="/test-file.drawio" data-test-resource-name="test-file.drawio" data-test-resource-type="file" class="oc-resource-name oc-text-truncate"><!----><span class="oc-resource-basename">test-file</span><span class="oc-resource-extension">.drawio</span></span></h2>
    <div>
      163 B,

      2 days ago
    </div>
  </div> <button aria-label="Click to mark this file as favorite" class="file_info__favorite oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
    <oc-icon-stub name="star" accessiblelabel="" type="span" size="medium" variation="passive" class="oc-star-dimm"></oc-icon-stub>
  </button>
</div>
`;

exports[`FileInfo FileInfo Components on { name: 'files-personal' } routes should show the fileinfo correctly 3`] = `
<div class="file_info">
  <oc-icon-stub name="folder" accessiblelabel="" type="span" size="large" variation="passive" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body">
    <h2 tabindex="-1"><span data-test-resource-path="/testfolder" data-test-resource-name="testfolder" data-test-resource-type="folder" class="oc-resource-name oc-text-truncate"><!----><span class="oc-resource-basename">testfolder</span>
      <!----></span>
    </h2>
    <div>
      163 B,

      2 days ago
    </div>
  </div> <button aria-label="Click to mark this file as favorite" class="file_info__favorite oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
    <oc-icon-stub name="star" accessiblelabel="" type="span" size="medium" variation="passive" class="oc-star-dimm"></oc-icon-stub>
  </button>
</div>
`;

exports[`FileInfo FileInfo Components on { name: 'files-public-list' } routes should show the fileinfo correctly 1`] = `
<div class="file_info">
  <oc-icon-stub name="file" accessiblelabel="" type="span" size="large" variation="passive" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body">
    <h2 tabindex="-1"><span data-test-resource-path="/testfile.txt" data-test-resource-name="testfile.txt" data-test-resource-type="file" class="oc-resource-name oc-text-truncate"><!----><span class="oc-resource-basename">testfile</span><span class="oc-resource-extension">.txt</span></span></h2>
    <div>
      163 B,

      2 days ago
    </div>
  </div> <button aria-label="Click to mark this file as favorite" class="file_info__favorite oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
    <oc-icon-stub name="star" accessiblelabel="" type="span" size="medium" variation="passive" class="oc-star-dimm"></oc-icon-stub>
  </button>
</div>
`;

exports[`FileInfo FileInfo Components on { name: 'files-public-list' } routes should show the fileinfo correctly 2`] = `
<div class="file_info">
  <oc-icon-stub name="file" accessiblelabel="" type="span" size="large" variation="passive" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body">
    <h2 tabindex="-1"><span data-test-resource-path="/test-file.drawio" data-test-resource-name="test-file.drawio" data-test-resource-type="file" class="oc-resource-name oc-text-truncate"><!----><span class="oc-resource-basename">test-file</span><span class="oc-resource-extension">.drawio</span></span></h2>
    <div>
      163 B,

      2 days ago
    </div>
  </div> <button aria-label="Click to mark this file as favorite" class="file_info__favorite oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
    <oc-icon-stub name="star" accessiblelabel="" type="span" size="medium" variation="passive" class="oc-star-dimm"></oc-icon-stub>
  </button>
</div>
`;

exports[`FileInfo FileInfo Components on { name: 'files-public-list' } routes should show the fileinfo correctly 3`] = `
<div class="file_info">
  <oc-icon-stub name="folder" accessiblelabel="" type="span" size="large" variation="passive" class="file_info__icon"></oc-icon-stub>
  <div class="file_info__body">
    <h2 tabindex="-1"><span data-test-resource-path="/testfolder" data-test-resource-name="testfolder" data-test-resource-type="folder" class="oc-resource-name oc-text-truncate"><!----><span class="oc-resource-basename">testfolder</span>
      <!----></span>
    </h2>
    <div>
      163 B,

      2 days ago
    </div>
  </div> <button aria-label="Click to mark this file as favorite" class="file_info__favorite oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
    <oc-icon-stub name="star" accessiblelabel="" type="span" size="medium" variation="passive" class="oc-star-dimm"></oc-icon-stub>
  </button>
</div>
`;
